### PR TITLE
Fix: Posix utils timespec overflow

### PR DESCRIPTION
**Pull request by abhidixi11**
Fixed potential UTILS_TimespecSubtract overflow
Updated UTILS_AbsoluteTimespecToTicks to get rid of call to posix function
Added unit tests

Description
-----------

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] My code is Linted.

### CODE REVIEW

File: tests/common/posix/aws_test_posix_utils.c [hunk 1]
```diff
@@ -255,3 +252,85 @@ TEST( Full_POSIX_UTILS, UTILS_strnlen )
 }
 
 /*-----------------------------------------------------------*/
+
+TEST( Full_POSIX_UTILS,  UTILS_TimespecCompare )
```
Comment (huguesBouvier) [inline tests/common/posix/aws_test_posix_utils.c]: Missing test case for xTimeX.tv_sec>xTimeY.tv_sec and the opposite

File: tests/common/posix/aws_test_posix_utils.c [hunk 2]
```diff
@@ -255,3 +252,85 @@ TEST( Full_POSIX_UTILS, UTILS_strnlen )
...
+    TEST_ASSERT_EQUAL_INT( EINVAL, UTILS_TimespecToTicks( &xTimeX, &xResult ) );
+    xTimeX.tv_nsec = -1;
+    TEST_ASSERT_EQUAL_INT( EINVAL, UTILS_TimespecToTicks( &xTimeX, &xResult ) );
+
+    /* Check return value and results with valid parameters*/
+    xTimeX.tv_sec = 1;
+    xTimeX.tv_nsec = NANOSECONDS_PER_TICK + 1;
+    TickType_t xExpectedVal = xTimeX.tv_sec * configTICK_RATE_HZ + 2;
+    TEST_ASSERT_EQUAL_INT( 0, UTILS_TimespecToTicks( &xTimeX, &xResult ) );
+    TEST_ASSERT_EQUAL_INT( ( int ) xResult, ( int ) xExpectedVal);
```
Comment (huguesBouvier) [inline tests/common/posix/aws_test_posix_utils.c]: Nitpick: Expect val is on the left.

File: lib/FreeRTOS-Plus-POSIX/source/FreeRTOS_POSIX_utils.c [hunk 3]
```diff
@@ -58,35 +58,33 @@ size_t UTILS_strnlen( const char * const pcString,
...
-            iStatus = errno;
-        }
-    }
-
-    /* Calculate the difference between the current time and pxAbsoluteTime. */
-    if( iStatus == 0 )
-    {
-        if( UTILS_TimespecSubtract( &xDifference, pxAbsoluteTime, &xCurrentTime ) != 0 )
+        iStatus = UTILS_TimespecSubtract( pxAbsoluteTime, pxCurrentTime, &xDifference );
+        if( iStatus == 1 )
```
[the remark this answers is not in the document]
Comment (abhidixi11) [reply lib/FreeRTOS-Plus-POSIX/source/FreeRTOS_POSIX_utils.c]: Negative number,  see utils.h for function description

File: lib/FreeRTOS-Plus-POSIX/source/FreeRTOS_POSIX_utils.c [hunk 4]
```diff
@@ -202,31 +200,103 @@ int UTILS_TimespecAddNanoseconds( struct timespec * const pxResult,
...
-    llResult64 = ( ( ( int64_t ) ( x->tv_sec ) * NANOSECONDS_PER_SECOND ) + ( int64_t ) ( x->tv_nsec ) )
-                 - ( ( ( int64_t ) ( y->tv_sec ) * NANOSECONDS_PER_SECOND ) + ( int64_t ) ( y->tv_nsec ) );
+    iCompareResult = UTILS_TimespecCompare( x, y );
 
-    /* Convert result to timespec. */
-    UTILS_NanosecondsToTimespec( llResult64, pxResult );
+    /* if x < y then result would be negative, return 1 */
+    if( iCompareResult == -1 )
+    {
+        return 1;
```
[the remark this answers is not in the document]
Comment (abhidixi11) [reply lib/FreeRTOS-Plus-POSIX/source/FreeRTOS_POSIX_utils.c]: That means the result would be negative number, so returning 1. Negative time indicates overflow.

File: lib/FreeRTOS-Plus-POSIX/source/FreeRTOS_POSIX_clock.c [hunk 5]
```diff
@@ -153,13 +154,19 @@ int clock_nanosleep( clockid_t clock_id,
         iStatus = EINVAL;
     }
 
+    /* Get current time */
+    if( clock_gettime( CLOCK_REALTIME, &xCurrentTime ) != 0 )
+    {
+        iStatus = errno;
```
[the remark this answers is not in the document]
Comment (abhidixi11) [reply lib/FreeRTOS-Plus-POSIX/source/FreeRTOS_POSIX_clock.c]: Good point, I'll fix in next commit.  I don't see errono implemented either, but it was part of original code.

File: tests/common/posix/aws_test_posix_utils.c [hunk 6]
```diff
@@ -135,58 +140,50 @@ TEST( Full_POSIX_UTILS, UTILS_TimespecSubtract )
...
-    TEST_ASSERT_EQUAL_INT( -1, xResult.tv_sec );
-    TEST_ASSERT_EQUAL_INT( 999999995, xResult.tv_nsec );
+    TEST_ASSERT_EQUAL_INT( 1, UTILS_TimespecSubtract( &x, &y, &xResult ) );
 
     /* Subtract non-compliant timespec where x.tv_nsec < 0. */
     x.tv_sec = y.tv_sec = 0;
     x.tv_nsec = -5; /* Equivalent to tv_sec = -1, tv_nsec = 999999995. */
     y.tv_nsec = 3;
-    TEST_ASSERT_EQUAL_INT( 1, UTILS_TimespecSubtract( &xResult, &x, &y ) );
-    TEST_ASSERT_EQUAL_INT( -1, xResult.tv_sec );
```
[the remark this answers is not in the document]
Comment (abhidixi11) [reply tests/common/posix/aws_test_posix_utils.c]: Please see comment above.

File: tests/common/posix/aws_test_posix_utils.c (hunk 2 above)
Comment (abhidixi11) [reply tests/common/posix/aws_test_posix_utils.c]: Good Point. Will make change in next commit

File: tests/common/posix/aws_test_posix_utils.c (hunk 1 above)
Comment (abhidixi11) [reply tests/common/posix/aws_test_posix_utils.c]: OK, will add

File: lib/FreeRTOS-Plus-POSIX/source/FreeRTOS_POSIX_mqueue.c [hunk 7]
```diff
@@ -172,15 +172,23 @@ static int prvCalculateTickTimeout( long lMessageQueueFlags,
...
             /* Check that the given timespec is valid. */
             if( UTILS_ValidateTimespec( pxAbsoluteTimeout ) == false )
             {
                 iStatus = EINVAL;
             }
 
+            /* Get current time */
+            if( clock_gettime( CLOCK_REALTIME, &xCurrentTime ) != 0 )
+            {
+                iStatus = errno;
```
[the remark this answers is not in the document]
Comment (abhidixi11) [reply lib/FreeRTOS-Plus-POSIX/source/FreeRTOS_POSIX_mqueue.c]: Will fix in the next commit.

**Comment by huguesBouvier** [comment]
Thank you!

**Pull request merged**